README.md
# HawkScan

[![PyPI version](https://d25lcipzij17d.cloudfront.net/badge.svg?id=py&type=6&v=2.2&x2=0)](https://pypi.org/project/hawkscan)
[![PyPI Statistics](https://img.shields.io/pypi/dm/hawkscan.svg)](https://pypistats.org/packages/hawkscan)
[![Twitter](https://img.shields.io/twitter/follow/c0dejump?label=c0dejump&style=social)](https://twitter.com/intent/follow?screen_name=c0dejump)


![alt tag](https://github.com/c0dejump/HawkScan/blob/master/static/logo_hawkscan.jpeg)

Security Tool for Reconnaissance and Information Gathering on a website. (python 3.x)

- [News](https://github.com/c0dejump/HawkScan/#News)
- [Installation](https://github.com/c0dejump/HawkScan/#Installation)
- [Special features](https://github.com/c0dejump/HawkScan/#Special-features)
- [TODO](https://github.com/c0dejump/HawkScan/#todo)
- [Usage](https://github.com/c0dejump/HawkScan/#usage)
- [Exemples](https://github.com/c0dejump/HawkScan/#exemples)
- [Thanks](https://github.com/c0dejump/HawkScan/#thanks)
- [Donations](https://github.com/c0dejump/HawkScan/#donations)
- [Tools used](https://github.com/c0dejump/HawkScan/#tools-used)
- [Wiki](https://github.com/c0dejump/HawkScan/wiki)

# News v2.x
    - Add proxy function
    - Redefining priorities/tasks
    - Let's debug certificate subdomains results
    - Display the current bypass number during scan ("CB:")
    - Easter egg for xmas :)
    - Option -nfs (not first step) to pass the first recon steps
    - Google CSE before scan
    - Creation of WIKI
    - Detecting potential path disclosure into html webpage
    - Detecting potential hidden directory
*(for more details go on CHANGELOG.md)* 
 
# Installation
``` 

       - git clone https://github.com/c0dejump/HawkScan.git && sudo python3 HawkScan/setup.py install
       
       - pip(3) install -r requirements.txt 
    
       - python3 -m pip install -r requirements.txt

``` 

# Special features

### Before scan
 - [x] Check header information
 - [x] Check DNS information
 - [x] Check Github
 - [x] CMS detection + version and vulns
 - [x] Check in waybackmachine
 - [x] Check if DataBase firebaseio existe and accessible
 - [x] Testing if it's possible scanning with "localhost" host
 - [x] Check Google Dork 
 - [x] Check Host IP
 - [x] Check backup domain name (ex: www.domain.com/domain.zip)
 - [x] Check socketio connection
 - [x] cse google search (buckets...)

### During - After scan
 - [x] Test backup/old file on all the files found (index.php.bak, index.php~ ...)
 - [x] Backup system (if the script stopped, it take again in same place)
 - [x] WAF detection and Response error to WAF + Testing bypass it
 - [x] Option --exclude to exclude page, code error, bytes
 - [x] Option rate-limit if app is unstable (--timesleep)
 - [x] Search S3 buckets in source code page
 - [x] Try differents bypass for 403/401 code error
 - [x] JS parsing and analysis (option --js)
 - [x] Auto resize relative to window
 - [x] Notify when scan completed (Only work on Linux)
 - [x] Multiple output format. Available formats: json, csv, txt
 - [x] Multiple website scanning
 - [x] Prefix filename (old_, copy of...)
 - [x] Detecting potential path disclosure into html webpage


# TODO 
**P1 is the most important**

 [WIP] Multiple exclude like: --exclude 403,1337b [P1] [In progress] (see [Exemples](https://github.com/c0dejump/HawkScan/#exemples))   
 [WIP] Anonymous routing through some proxy (http/s proxy list) [P1] [In progress]
 [WIP] Re-build resport scan [P1]
 [WIP] HExHTTP replace "header information" before scan
 - [ ] asyncio instead of threading ? [PX]
 - [ ] Add crt.sh to check potential hidden subdomain (with letdebug module ?) [PX]
 - [ ] Push results into DB [P2]
 - [ ] If re-scan a website with an existing folder, just doing a diff btw the scan to the folder (like) // interesting ? [P2]
 - [ ] Pre-run to check the waf sensitive (by proxy with 40 threads for exemple) // add proxy funtion [P2]
 - [ ] Check source code and verify leak or sensitive data in Github // Other tool ? [P3]
 - [ ] Scan API endpoints/informations leaks [P3]

# Usage
  
```
     
    usage: hawkscan.py [-h] [-u URL] [-f FILE_URL] [-t THREAD] [--exclude EXCLUDE [EXCLUDE ...]] [--auto] [--update] [-w WORDLIST] [-b [BACKUP ...]] [-p PREFIX] [-H HEADER_] [-a USER_AGENT] [--redirect] [--auth AUTH] [--timesleep TS] [--proxie PROXIE] [-r] [-s SUBDOMAINS] [--js] [--nfs] [--ffs] [--notify] [-o OUTPUT] [-of OUTPUT_TYPE]    
 
```

``` 
> General:
    -u URL                URL to scan [required]
    -f FILE_URL           file with multiple URLs to scan
    -t THREAD             Number of threads to use for URL Fuzzing. Default: 30
    --exclude EXCLUDE [EXCLUDE ...] Exclude page, response code, response size. (Exemples: --exclude 500,337b)   
    --auto                Automatic threads depending response to website. Max: 30
    --update              For automatic update
    --lightmode           For a just simple fuzzing 1 request per second & a new session for each request


> Wordlist Settings:
    -w WORDLIST           Wordlist used for Fuzzing the desired webite. Default: dichawk.txt     
    -b                    Adding prefix/suffix backup extensions during the scan. (Exemples: exemple.com/~ex/, exemple.com/ex.php.bak...) /!\ beware, take more longer
    -p PREFIX             Add prefix in wordlist to scan

> Request Settings:             
    -H HEADER_            Modify header. (Exemple: -H "cookie: test")    
    -a USER_AGENT         Choice user-agent. Default: Random    
    --redirect            For scan with redirect response (301/302)      
    --auth AUTH           HTTP authentification. (Exemples: --auth admin:admin)               
    --timesleep TS        To define a timesleep/rate-limit if app is unstable during scan.

> Tips:            
    -r                    Recursive dir/files      
    -s SUBDOMAINS         Subdomain tester         
    --js                  For try to found keys, token, sensitive endpoints... in the javascript page
    --nfs                 Not the first step of scan during the first running (waf, vhosts, wayback etc...)    
    --ffs                 Force the first step of scan during the first running (waf, vhosts, wayback etc...)              
    --notify              For receveid notify when the scan finished (only work on linux)

> Export Settings:                    
    -o OUTPUT             Output to site_scan.txt (default in website directory)     
    -of OUTPUT_TYPE       Output file format. Available formats: json, csv, txt           
```

# Exemples

```
    //Basic
     python hawkscan.py -u https://www.exemple.com/

    //With specific dico
     python hawkscan.py -u https://www.exemple.com/ -w dico_extra.txt

    //with 30 threads
     python hawkscan.py -u https://www.exemple.com/ -t 30

    //With backup files scan
     python hawkscan.py -u https://www.exemple.com/ -b

    //With an exclude page
     python hawkscan.py -u https://www.exemple.com/ --exclude profile.php

    //With an exclude response code
     python hawkscan.py -u https://www.exemple.com/ --exclude 403

    //With an exclude bytes number
     python hawkscan.py -u https://www.exemple.com/ --exclude 1337b 

    //With two excludes type
     python hawkscan.py -u https://www.exemple.com/ --exclude 1337b,403

```

# Thanks
Layno (https://github.com/Clayno/) [Technical helper]      
Sanguinarius (https://twitter.com/sanguinarius_Bt) [Technical helper]  
Jamb0n69 (https://twitter.com/jamb0n69) [Technical helper]           
Cyber_Ph4ntoM (https://twitter.com/__PH4NTOM__) [Beta tester & Logo Graphist]


# Donations

https://www.paypal.me/c0dejump

Or if you want to offer me a coffee :)

https://ko-fi.com/c0dejump


## Tools used

This script use "WafW00f" to detect the WAF in the first step (https://github.com/EnableSecurity/wafw00f)

This script use "Sublist3r" to scan subdomains (https://github.com/aboul3la/Sublist3r)


CHANGELOG.md
Changelog:
----------

- 2.6
---------
	Updated: dichawk endpoints
	New: lightmode option
---------

- 2.5
---------
	Updated: half of the code has been refactored/reorganized
	Updated: dichawk endpoints
	Updated: JS sensitives endpoints
	Updated: Exclude option (referer to Readme)
	Fixed: ANy error with exclude option
	New: Download all js link with specific keyword found
---------

- 2.3.1
---------
	Updated: dichawk endpoints
	Updated: JS sensitives endpoints
---------

- 2.3
---------
	Started: Add proxy function (--proxy proxy.lst) [In progress]
	New: Adding path disclosure recon
	New: Detecting potential hidden directory
	Updated: dichawk endpoints
---------

- 2.2
---------
	New: Wiki created !
	Fixed: any bugs and redesign titles pre-scan
	Changed: waybacktool module deleted by a personal module (less slow)
---------

- 2.1
---------
	New: Option -nfs (not first step) to pass the first recon steps
	Fixed: Any bug with the download file and bypass forbidden when differents options
	New: Google cse search (buckets...)
	New: Add LICENSE & PyPI version and stats
---------

- 2.0
---------
	Redefining priorities/tasks
	New: Let's debug certificate subdomains results
	New: Display the current bypass number during scan ("CB:")
	New: Easter egg for xmas :)
	Updated: Fix any bugs

---------

- 1.9.9
---------
	New: Cloudflare protection detection in live
	Updated: Bugs correction with backup extension scan
---------

- 1.9.8
---------
	Updated: New ways to bypass forbidden (Thanks @yunem_se !)
	Updated: New socketio endpoints
	Updated: New words in dichawk.txt
---------

- 1.9.7
---------
	Updated: New logo by @__PH4NTOM__!
---------

- 1.9.6
---------
	Fixed:	 Any bugs
	Updated: Little style modifications
---------

- 1.9.5
---------
	Fixed: A pass on the source code, more speedy
---------

- 1.9.4
---------
	Added: 	Function "vim_backup" to test backup vim during scan when any -b option
---------

- 1.9.3
---------
	Updated: New banner
	Fixed:	 Multiple website with file which contain url

- 1.9.2:
---------
	Updated: code review & optimisation
	Updated: Multiple new paths/words in dichawk.txt
---------

- 1.9.1
---------
	Added: 	 Option "-f" for scanning multiple website to one time. Ex: ```-f urls_file.txt```
	Updated: Clean code & directory
---------

- 1.9
---------
	Fixed: Fixed percentage & line count bug during scan
	Added: Display errors number in live during scan 
---------

- 1.8.8
---------
	Added: Output file format function. Available formats: json, csv, txt. Ex: ```-of json```, ```-o /tmp/Target -of csv```
---------

- 1.8.7
---------
	Fixed: Reduction of false positives number
	Added: Header of Hawkscan when your typing "head hawkscan.py"
---------

- 1.8.6
---------
	Fixed:	 Any bugs: Thread modification, header parameters, bypass forbidden & any others...
	Added: 	 google module in requirements/setup
	Updated: Deleted degoogle modules/script, google dork works now with the "googlesearch" module
	Updated: A little style modification
	Updated: Default thread now 30
---------

- 1.8.5:
---------
	Added: 	A new restriction bypass feature, that test "post,put,patch & option" requests methods on a restriction page
	Fixed: 	The little style display problems
---------

- 1.8.4
---------
	Fixed: 	Better display of live lines
	Added: 	A new file to manage the modules to launches
---------

- 1.8.3
---------
	Fixed: 	Bug in socketio module
	Fixed: 	Add size bytes during th error scan
	Added: 	Words in wordlist
---------

- 1.8.2
---------
	Updated: New logo made by Cyber_Ph4ntoM
	Updated:  Code review
	Updated:  Add multiple words in dichawk.txt
---------

- 1.8.1
---------
	Updated:  Style refont
	Fixed:    Mutliple bugs
---------

- 1.8
---------
	Update:   you can choose your backup file when you scan, EX:
		-b: default all backup file
		-b .bck, .old, .bak: just these backups

- 1.7.9
---------
	Updated:  dico.txt → dichawk.txt (dico.txt it was to simple for a personal dictionary :)
	Fixed:    Bug on parsing JS
---------

- 1.7.8
---------
	Fixed:    Bug on the exclude function
	Fixed:    Bug on the bypass forbidden function
	Added:    News header value in bypass forbidden function
---------

- 1.7.7
---------
	Updated:  Rrefont helping style
	Added:    Notify when scan completed (Only work on Linux)
---------

- 1.7.6
---------
	Added: New function added: check_socketio(), to check the websocket requests during the first step. That can potentially leak any informations (path, message, users...). Adding too in JS verification to check if any endpoint look like socketio.
	Fixed: Reducted false positive number
---------

- 1.7.5
---------
	Deleted: WhoIs function (useless)
	Updated: Style refont
---------

- 1.7.4
---------
	Updated: Dockerfile
	Added: A resume of commands (url, threads...) during the begin scan
---------

- 1.7.3
---------
	Updated: setup.py
---------

- 1.7.2
---------
	Updated: Add new content in google dork, dico and javascript recon
	Updated: Real setup.py :)
---------

- 1.7.1
---------
	Fixed: Any bugs
	Modified: Raw output, modification for any integration
---------

- 1.7
---------
	Added: Function "check_backup_domain" added, test before start "domain.zip/rar etc.."
	Added: New option (-ffs) to force the first step of scan during the first running (waf, vhosts, wayback etc...)
---------

- 1.6.9
---------
	Added: multiple excludes (only number or bytes number), exemple: --exclude 403,500 // --exclude 1337b,500...
---------

- 1.6.8
---------
	Fixed: Bugs
	Modified: function "check_ip" which check if the IP o the website was different by domain name => Now: check_vhost
---------

- 1.6.7
---------
	Added: Option --auth for HTTP authentification. Exemple --auth admin:admin
	Fixed: Bugs
---------

- 1.6.6

modules/before_run.py
import requests
from bs4 import BeautifulSoup
import json
import sys, re, os
import ssl
import socket
import traceback
from requests.exceptions import Timeout
import time
# External
from config import PLUS, WARNING, INFO, LESS, LINE, FORBI, BACK, INFO_MOD

requests.packages.urllib3.disable_warnings(requests.packages.urllib3.exceptions.InsecureRequestWarning)

class before_start:
    

    def get_header(self, url, directory):
        """Get header of website (cookie, link, etc...)"""
        r = requests.get(url, allow_redirects=False, verify=False)
        print("\033[36m HEADER\033[0m")
        print(LINE)
        print("  {} \n".format(r.headers).replace(',','\n'))
        print(LINE)
        with open(directory + '/header.csv', 'w+') as file:
            file.write(str(r.headers).replace(',','\n'))


    def gitpast(self, url):
        """
        Github: check github informations
        """
        print("\033[36m Github \033[0m")
        print(LINE)
        url = url.split(".")[1] if "www" in url else url.split("/")[2]
        url = "{}".format(url)
        print("search: {}\n".format(url))
        types = ["Commits", "Issues", "Repositories", "Topics", "Wikis", "Users", "Code"]
        try:
            for t in types:
                github = "https://github.com/search?q={}&type={}".format(url, t)
                req = requests.get(github, verify=False)
                soup = BeautifulSoup(req.text, "html.parser")
                search = soup.find('a', {"class":"menu-item selected"})
                if search:
                    for s in search.find("span"):
                        print("  {}{}: {}".format(INFO_MOD, t, s))
                else:
                    print("  {}{}: not found".format(INFO_MOD, t))
        except:
            print("{}You need connection to check the github".format(WARNING))
        print("\n" + LINE)


    def get_dns(self, url, directory):
        """Get DNS informations"""
        port = 0
        print("\033[36m DNS information \033[0m")
        print(LINE)
        try:
            if "https" in url:
                url = url.replace('https://','').replace('/','')
                port = 443
            else:
                url = url.replace('http://','').replace('/','')
                port = 80
            context = ssl.create_default_context()
            conn = context.wrap_socket(socket.socket(socket.AF_INET), server_hostname=url)
            conn.connect((url, port))
            cert = conn.getpeercert()
            print(" \u251c Organization: {}".format(cert['subject']))
            print(" \u251c DNS: {}".format(cert['subjectAltName']))
            print(" \u251c SerialNumber: {}".format(cert['serialNumber']))
            conn.close()
            with open(directory + '/dns_info.csv', 'w+') as file:
                file.write(str(cert).replace(',','\n').replace('((','').replace('))',''))
        except:
            erreur = sys.exc_info()
            typerr = u"%s" % (erreur[0])
            typerr = typerr[typerr.find("'")+1:typerr.rfind("'")]
            print(typerr)
            msgerr = u"%s" % (erreur[1])
            print(msgerr + "\n")
        print(LINE)


    def letsdebug(self, url):
        """
        letsdebug: Get certificate of the website and potentialy found old certificate with old subdomain
        """
        print("\033[36m Let's Debug information \033[0m")
        print(LINE)
        try:
            wait_finish = True
            list_result = []
            string_result = ""
            domain = ".".join(url.split("/")[2].split(".")[1:]) if len(url.split("/")[2].split(".")) == 3 else ".".join(url.split("/")[2].split(".")[0:])
            url_ld = "https://letsdebug.net/"
            print(" {} {}".format(INFO_MOD, domain))
            datas = {"domain":domain,"method":"http-01"}
            req = requests.post(url_ld, data=datas, allow_redirects=True, verify=False)
            url_debug = "{}?debug=y".format(req.url)
            while wait_finish:
                res = requests.get(url_debug, verify=False)
                if "please wait" in res.text:
                    time.sleep(1)
                else:
                    wait_finish = False
            soup = BeautifulSoup(res.text, "html.parser")
            search = soup.find('div', {"id":"RateLimit-Debug"})
            if search:
                for s in search:
                        if s != None and s != "\n":
                            string_result += str(s)
                result = re.findall(r'\[.*?\]', string_result)
                for r in result:
                    r = r.replace("[","").replace("]","")
                    if r not in list_result:
                        list_result.append(r)
                for rl in list_result:
                    print(" {} {}".format(INFO_MOD, rl))
            else:
                print(" {} Nothing certificate subdomain found".format(INFO_MOD))
        except:
            pass
        print(LINE)


    def firebaseio(self, url):
        """
        Firebaseio: To check db firebaseio
        ex: --firebase facebook
        """
        get_domain = url.split("/")[2]
        parse_domain = get_domain.split(".")
        if not "www" in get_domain:
            dire = "{}-{}".format(parse_domain[0], parse_domain[1]) if len(parse_domain) > 2 else "{}".format(parse_domain[0])
        else:
            dire = "{}".format(parse_domain[1])
        print("\033[36m Firebaseio \033[0m")
        print(LINE)
        url = 'https://{}.firebaseio.com/.json'.format(dire.split(".")[0])
        print(" Target: {}\n".format(url))
        try:
            r = requests.get(url, verify=False).json()
            if 'error' in r.keys():
                if r['error'] == 'Permission denied':
                    print(" {}{} seems to be protected".format(FORBI, url)) #successfully protected
                elif r['error'] == '404 Not Found':
                    print(" {}{} not found".format(LESS, url)) #doesn't exist
                elif "Firebase error." in r['error']:
                    print(" {}{} Firebase error. Please ensure that you spelled the name of your Firebase correctly ".format(WARNING, url))
            else:
                print(" {}{} seems to be vulnerable !".format(PLUS, url)) #vulnerable
        except AttributeError:
            '''
            Some DBs may just return null
            '''
            print("{} null return".format(INFO))
        except:
            print("Error with the requests, please do a manual check")
            pass
        print(LINE)
        

    def wayback_check(self, url, directory):
        """
        Wayback_check:
        Check in a wayback machine to found old file on the website or other things...
        """
        print("\033[36m Wayback \033[0m")
        print(LINE)
        print(url + "\n")
        url_wayb = "http://web.archive.org/cdx/search?url=*.{}/*&output=list&fl=original,statuscode&collapse=urlkey&filter=!statuscode:404".format(url)
        try:
            req_way = requests.get(url_wayb, verify=False)
            urls_w = req_way.text.split("\n")
            #print(urls_w)
            for url_w in urls_w:
                try:
                    status_c = url_w.split(" ")[1]
                    url_ext = url_w.split(" ")[0]
                    if "css" not in url_ext and "jpg" not in url_ext and "png" not in url_ext and \
                     "jpeg" not in url_ext and "svg" not in url_ext and "JPG" not in url_ext and "gif" not in url_ext and \
                     "mov" not in url_ext and "mp4" not in url_ext and "woff" not in url_ext and "ttf" not in url_ext and \
                     "eot" not in url_ext:
                        if status_c in ['200', '403', '401', '301', '302']:
                            with open("{}/wayback.txt".format(directory), "a+") as w:
                                w.write("{} {}\n".format(url_ext, status_c))
                        if status_c == "200":
                            print("  \033[32m\u251c\033[0m {} \033[32m{}".format(url_ext, status_c))
                        elif status_c in ['403', '401']:
                            print("  \033[31m\u251c\033[0m {} \033[31m{}".format(url_ext, status_c))
                        elif status_c in ['301', '302']:
                            print("  \033[33m\u251c\033[0m {} \033[33m{}".format(url_ext, status_c))
                        else:
                            pass
                except:
                    pass

        except:
            #traceback.print_exc() 
            print(" {} An error occurred please check manually ".format(LESS))
        print(LINE)


    def check_localhost(self, url):
        """
        Check_localhost: Function which try automatically if it's possible scanning with "localhost" host for discovery other files/directories
        """
        list_test = ["127.0.0.1", "localhost"]
        localhost = False
        print("\033[36m Localhost host \033[0m")
        print(LINE)
        for lt in list_test:
            header = {"Host": lt}
            try:
                req = requests.get(url, headers=header, verify=False, timeout=10)
                if req.status_code == 200:
                    print(" \033[32m\u251c\033[0m You can potentialy try bf directories with this option '-H \"Host:{}\"' ".format(lt))
                    localhost = True
                else:
                    pass
            except:
                pass
        if not localhost:
            print(" {} Not seem possible to scan with localhost host".format(LESS))
        print(LINE)


    def check_vhost(self, domain, url):
        """
        check_ip:
        Check the host ip if this webpage is different or not
        """
        print("\033[36m Vhosts misconfiguration \033[0m")
        print(LINE)
        try:
            req_index = requests.get(url, verify=False, timeout=10)
            len_index = len(req_index.content)
            retrieve_ip = False
            dom = socket.gethostbyname(domain)
            ips = ["https://{}/".format(dom), "http://{}/".format(dom), "http://www2.{}/".format(domain), "http://www3.{}/".format(domain), "https://www2.{}/".format(domain),
            "https://www3.{}/".format(domain)]
            for ip in ips:
                try:
                    req_ip = requests.get(ip, verify=False, timeout=10)
                    if req_ip.status_code not in [404, 403, 425, 503, 500, 400] and len(req_ip.content) != len_index:
                        retrieve_ip = True
                        print(" \033[32m\u251c\033[0m The host IP seem to be different, check it: {} ".format(ip))
                except:
                    #print(" \033[33m\u251c\033[0m The host IP have a problem, check it manualy please: {} ".format(ip))
                    pass
            if not retrieve_ip:
                print(" {} IPs do not appear to be different from the host".format(LESS))
            print(LINE)
        except:
            pass


    def check_backup_domain(self, domain, url):
        """
        check_backup_domain:
        Check the backup domain, like exemple.com/exemple.zip
        """
        print("\033[36m Domain backup \033[0m")
        print(LINE)
        backup_dn_ext = ["zip", "rar", "iso", "tar", "gz", "tgz", "tar.gz", "7z", "jar", "sql.gz"]
        found_bdn = False
        len_response = 0
        try:
            req_index = requests.get(url, verify=False, timeout=10)
            len_index = len(req_index.content)
            domain = domain.split('.')[1] if len(domain.split('.')) > 2 else domain.split('.')[0]
            print(" {}List of backup extension for domain {}: {}\nExemple: {}{}.zip\n".format(INFO, domain, backup_dn_ext, url, domain.split('.')[0]))
            for bdn in backup_dn_ext:
                url_dn_ext = "{}{}.{}".format(url, domain.split('.')[0], bdn)
                try:
                    req_dn_ext = requests.get(url_dn_ext, verify=False, timeout=10)
                    if req_dn_ext.status_code not in [404, 403, 401, 500, 400, 425] and len(req_dn_ext.content) not in range(len_index - 10, len_index + 10):
                        if len(req_dn_ext.content) not in range(len_response - 10, len_response + 10):
                            print(" {} {} found ({}b)".format(PLUS, url_dn_ext, len(req_dn_ext.text)))
                            len_response = len(req_dn_ext.content)
                            found_bdn = True
                except:
                    pass
        except:
            pass
        if not found_bdn:
            print(" {} No backup domain name found".format(LESS))
        print(LINE)



    def test_timeout(self, url, first=False):
        """
        Test_timeout: just a little function for test if the connection is good or not
        """
        try:
            req_timeout = requests.get(url, timeout=30, verify=False)
        except Timeout:
            print("{}Service potentialy Unavailable, The site web seem unavailable please wait...\n".format(WARNING))
            if first:
                next_step = input("Do you want continue ?: [y:N] ")
                if next_step in ["y", "Y"]:
                    pass
                else:
                    sys.exit()
            else:
                time.sleep(180)
        except requests.exceptions.ConnectionError:
            pass